bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+            State::Learning => "learning",
+            State::Learned { .. } => "learned",
+        }
+    }
+}
+
+/// A response to an Fsm API request
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ApiOutput {
+    /// This peer has been initialized
+    ///
+    /// The caller *must* persist the state
+    PeerInitialized,
+
+    /// Rack initialization has completed. This node was the coordinator.
+    RackInitComplete,
+
+    /// A `RackSecret` was reconstructed
+    RackSecret { request_id: Uuid, secret: RackSecret },
+
+    /// An extra share has been distributed to a learning peer
+    ///
+    /// The caller *must* persist the state
+    ShareDistributedToLearner,
+
+    /// This peer Learned its share
+    ///
+    /// The caller *must* persist the state
+    LearningCompleted,
+}
+
+/// An error returned from an Fsm API request
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum ApiError {
+    #[error("already initialized")]
+    AlreadyInitialized,
+
+    #[error("not yet initialized")]
+    NotInitialized,
+
+    #[error("cannot retrieve or distribute shares while learning")]
+    StillLearning,
+
+    #[error("rack init timeout: unacked_peers: {unacked_peers:?}")]
+    RackInitTimeout { unacked_peers: BTreeSet<Baseboard> },
+
+    #[error("rack init falied: trust quorum error: {0:?}")]
+    RackInitFailed(TrustQuorumError),
+
+    #[error("rack secret load timeout")]
+    RackSecretLoadTimeout,
+
+    #[error("share from {from} has invalid sha3_256 digest")]
+    InvalidShare { from: Baseboard },
+
+    #[error("critical: failed to reconstruct rack secret with valid shares")]
+    FailedToReconstructRackSecret,
+
+    #[error("critical: failed to decrypt extra shares")]
+    FailedToDecryptExtraShares,
+
+    #[error("unexpected response ({msg}) from ({from}) in state ({state}) with request_id ({request_id})")]
+    UnexpectedResponse {
+        from: Baseboard,
+        state: &'static str,
+        request_id: Uuid,
+        msg: &'static str,
+    },
+
+    #[error("error response received from ({from}) in state ({state}) with request_id ({request_id}): {error:?}")]
+    ErrorResponseReceived {
+        from: Baseboard,
+        state: &'static str,
+        request_id: Uuid,
+        error: MsgError,
+    },
+}
+
+pub struct Fsm {
+    /// The current state of this peer
+    state: State,
+    /// Unique ID of this peer
+    id: Baseboard,
+
+    /// User provided configuration
+    config: Config,
+
+    /// Unique IDs of connected peers
+    connected_peers: BTreeSet<Baseboard>,
+
+    /// Manage all trackable requests
+    request_manager: RequestManager,
+
+    /// Envelopes not managed by the `RequestManager`
+    /// These are all envelopes containing `Response` messages
+    envelopes: Vec<Envelope>,
+
+    /// We keep track of whether the rack failed to initialize. If this happens
+    /// the coordinator should return this error on every new API request.
+    rack_init_error: Option<(Uuid, ApiError)>,
```

Comment:
This suggests that we might want a top-level enum, e.g.
```rust
enum Fsm {
    Running(RunningFsm),
    Failed(Uuid, ApiError),
}
```
(where `struct RunningFsm` has everything in the current `Fsm` struct)

Not sure how much that would mess up, though.

Reply:
Seems like overkill to me at this point.

Reply:
So I took a quick stab at this. I believe doing so means that all `Fsm` methods need to now operate at the top level and possibly be fallible or we need all internal fields in the current `Fsm` moved to a separate struct and both variants would need access to that. Then we'd have to duplicate all methods and/or add a trait with a default implementation. This will all add a lot of code for specious value.